packages/simulation/src/convert.ts

```diff
@@ -0,0 +1,647 @@
+    const currentState = stack.pop() as StateID
+    for (const transition of dfaGraph.transitions) {
+      if (transition.from === currentState) {
+        // If the transition leads to a new state, add it to the reachable states
+        if (!reachableStates.has(transition.to)) {
+          reachableStates.add(transition.to)
+          stack.push(transition.to)
+        }
+      }
+    }
+  }
+  // Remove any states that are not in the set of reachable states
+  dfaGraph.states = dfaGraph.states.filter((state) => reachableStates.has(state.id))
+  // Remove any transitions that are not connected to reachable states
+  dfaGraph.transitions = dfaGraph.transitions.filter((transition) => reachableStates.has(transition.from) && reachableStates.has(transition.to))
+  return dfaGraph
+}
+
+// This will check to ensure that the graph passed in has valid states/transitions before continuing
+export const statesAndTransitionsPresent = (nfaGraph: FSAGraphIn): boolean => {
+  if (nfaGraph.states.length === 0 || nfaGraph.transitions.length === 0) {
+    return false
+  } else {
+    return true
+  }
+}
+
+// This will check to ensure that the graph passed in has a final state before continuing
+export const finalStateIsPresent = (nfaGraphStates: FSAState[]): boolean => {
+  const numberOfNFAStates: number = nfaGraphStates.length
+  let doesFinalExist: boolean = false
+  for (let curElem = 0; curElem < numberOfNFAStates; curElem++) {
+    if (nfaGraphStates[curElem].isFinal) {
+      doesFinalExist = true
+    }
+  }
+  if (!doesFinalExist) {
+    return false
+  } else {
+    return true
+  }
+}
+
+// This will check to ensure that the graph passed in has an initial state before continuing
+export const initialStateIsPresent = (nfaGraphInitialState: StateID): boolean => {
+  if (nfaGraphInitialState == null) {
+    return false
+  } else {
+    return true
+  }
+}
+
+// This will check if the merged state is different from the current state's merged state
+function arraysEqual (arr1, arr2) {
+  if (arr1.length !== arr2.length) {
+    return false
+  }
+  for (let i = 0; i < arr1.length; i++) {
+    if (arr1[i] !== arr2[i]) {
+      return false
+    }
+  }
+  return true
+}
+
+// This will check that a given states transitions contain all the symbols
+function hasAllSymbols (transitions, symbols) {
+  for (const symbol of symbols) {
+    let symbolFound = false
+    for (const [, readSymbol] of transitions) {
+      if (readSymbol === symbol) {
+        symbolFound = true
+        break
+      }
+    }
+    if (!symbolFound) {
+      return false
+    }
+  }
+  return true
+}
+
+// This function will handle the logic for combining states when necessary
+function combineStates (initialTransitionTable: {[key: StateID]: [StateID, ReadSymbol][]}, symbolsArray: ReadSymbol[], curTransitionElem: number, symbolToStatesMap: {[key: ReadSymbol]: [StateID, StateID][]}, symbol: ReadSymbol): [string, { symbol: ReadSymbol, tostates: StateID[] }[]] {
```

Comment:
Make a type alias for the `initialTransitionTable` at the start of the file and use that
```ts
type TransitionTable = Record<StateID, [StateID, ReadSymbol][]>
```
I also notice a lot of loops to find states, might it be better to have it be `Record<StateID, Record<StateID, ReadSymbol>>` or is the algorithm easier this way?